systemd_ctypes/bus.py

```diff
@@ -320,124 +327,467 @@ def add_match(self, rule, handler):
+                counter = bus.Interface.Property('i', value=0)
+                @counter.setter
+                def set_counter(self, value):
+                    # we got a request to set the counter from D-Bus
+                    self.counter = value
+
+        In all cases, the first (and only mandatory) argument to the
+        constructor is the D-Bus type of the property.
+
+        Your getter and setter functions can be provided by kwarg to the
+        constructor.  You can also give a name= kwarg to override the default
+        name conversion scheme.
+        """
+        _category = 'properties'
+
+        _getter: Optional[Callable[[Any], Any]]
+        _setter: Optional[Callable[[Any, Any], None]]
+        _type_string: str
+        _value: Any
+
+        def __init__(self, type_string: str,
+                     value: Any = None,
+                     name: Optional[str] = None,
+                     getter: Optional[Callable[[Any], Any]] = None,
+                     setter: Optional[Callable[[Any, Any], None]] = None):
+            assert value is None or getter is None, 'A property cannot have both a value and a getter'
+
+            super().__init__(name=name)
+            self._getter = getter
+            self._setter = setter
+            self._type_string = type_string
+            self._value = value
+
+        def _describe(self) -> Dict[str, Any]:
+            return {'type': self._type_string, 'flags': 'r' if self._setter is None else 'w'}
+
+        def __get__(self, obj: object, cls: Optional[type] = None) -> Any:
+            if obj is None:
+                return self
+            if self._getter is not None:
+                return self._getter.__get__(obj, cls)()
+            elif self._value is not None:
+                if obj._dbus_property_values is not None:
+                    return obj._dbus_property_values.get(self._name, self._value)
+                else:
+                    return self._value
+            else:
+                raise AttributeError(f"'{obj.__class__.__name__}' property '{self._python_name}' "
+                                     f"was not properly initialised: use either the 'value=' kwarg or "
+                                     f"the @'{self._python_name}.getter' decorator")
+
+        def __set__(self, obj: Any, value: Any) -> None:
+            if self._getter is not None:
+                raise AttributeError(f"'{obj.__class__.__name__}' property '{self._python_name}' is read-only")
+            if obj._dbus_property_values is None:
+                obj._dbus_property_values = {}
+            obj._dbus_property_values[self._name] = value
+            obj.properties_changed(self._interface or obj._dbus_interface,
+                                   {self._name: {"t": self._type_string, "v": value}},
+                                   [])
+
+        def to_dbus(self, obj):
+            return {"t": self._type_string, "v": self.__get__(obj)}
```

Comment:
I start to think we could have a more abstract type for variants. Even cockpit.js has that, but nobody uses it... which might be a hint, hmm.

Reply:
Something like

```
class Variant:
  def __init__(self, type, value):
    self.t = type
    self.v = value
```

Reply:
My "new crack" approach introduces this type, indeed.  I want to focus more on Cockpit again before merging that, though.